dxr/lines.py

```diff
@@ -35,6 +37,167 @@ def __repr__(self):
+
+        """
+        self.tree = tree
+        self.menu_data = menu_data
+        self.hover = hover
+        self.qualname_hash = hash(qualname) if qualname else qualname_hash
+
+    def es(self):
+        """Return a serialization of myself to store in elasticsearch."""
+        ret = {'plugin': self.plugin,
+               'id': self.id,
+               # Smash the data into a string, because it will have a
+               # different schema from subclass to subclass, and ES will freak
+               # out:
+               'menu_data': json.dumps(self.menu_data)}
+        if self.hover:
+            ret['hover'] = self.hover
+        if self.qualname_hash is not None:  # could be 0
+            ret['qualname_hash'] = self.qualname_hash
+        return ret
+
+    @staticmethod
+    def es_to_triple(es_data, tree):
+        """Convert ES-dwelling ref representation to a (start, end,
+        :class:`~dxr.lines.Ref` subclass) triple.
+
+        Return a subclass of Ref, chosen according to the ES data. Into its
+        attributes "menu_data", "hover" and "qualname_hash", copy the ES
+        properties of the same names, JSON-decoding "menu_data" first.
+
+        :arg es_data: An item from the array under the 'refs' key of an ES LINE
+            document
+        :arg tree: The :class:`~dxr.config.TreeConfig` representing the tree
+            from which the ``es_data`` was pulled
+
+        """
+        def ref_class(plugin, id):
+            """Return the subclass of Ref identified by a combination of
+            plugin and class ID."""
+            plugins = all_plugins()
+            try:
+                return plugins[plugin].refs[id]
+            except KeyError:
+                warn('Ref subclass from plugin %s with ID %s was referenced '
+                     'in the index but not found in the current '
+                     'implementation. Ignored.' % (plugin, id))
+
+        payload = es_data['payload']
+        cls = ref_class(payload['plugin'], payload['id'])
+        return (es_data['start'],
+                es_data['end'],
+                cls(tree,
+                    json.loads(payload['menu_data']),
+                    hover=payload.get('hover'),
+                    qualname_hash=payload.get('qualname_hash')))
+
+    def menu_items(self):
+        """Return an iterable of menu items to be attached to a ref.
+
+        Return an iterable of dicts of this form::
+
+            {
+                html: the HTML to be used as the menu item itself
+                href: the URL to visit when the menu item is chosen
+                title: the tooltip text given on hovering over the menu item
+                icon: the icon to show next to the menu item: the name of a PNG
+                    from the ``icons`` folder, without the .png extension
+            }
+
+        Typically, this pulls data out of ``self.menu_data``.
+
+        """
+        raise NotImplementedError
+
+    def opener(self):
```

Comment:
I think a docstring would be nice here, perhaps to describe how we put menu items into the HTML so we can pull them out on clicks.